src/staking.rs

```diff
@@ -1052,1512 +1053,1487 @@ mod test {
+        )?)?)
+    }
 
-        router
-            .staking
-            .setup(
-                &mut store,
-                StakingInfo {
-                    bonded_denom: BONDED_DENOM.to_string(),
-                    unbonding_time: 60,
-                    apr: Decimal::percent(apr),
-                },
-            )
-            .unwrap();
+    /// Executes distribution message.
+    fn execute_distr(
+        env: &mut TestEnv,
+        sender: Addr,
+        msg: DistributionMsg,
+    ) -> AnyResult<AppResponse> {
+        env.router.distribution.execute(
+            &env.api,
+            &mut env.storage,
+            &env.router,
+            &env.block,
+            sender,
+            msg,
+        )
+    }
 
-        // add validator
-        let valoper1 = Validator::new(
-            validator_addr.to_string(),
-            Decimal::percent(validator_commission),
-            Decimal::percent(100),
-            Decimal::percent(1),
-        );
-        router
-            .staking
-            .add_validator(&api, &mut store, &block, valoper1)
+    /// Executes bank query.
+    fn query_bank<T: DeserializeOwned>(env: &TestEnv, msg: BankQuery) -> AnyResult<T> {
+        Ok(from_json(env.router.bank.query(
+            &env.api,
+            &env.storage,
+            &env.router.querier(&env.api, &env.storage, &env.block),
+            &env.block,
+            msg,
+        )?)?)
+    }
+
+    /// Initializes balance for specified address in staking denominator.
+    fn init_balance(env: &mut TestEnv, address: &Addr, amount: u128) {
+        init_balance_denom(env, address, amount, BONDED_DENOM);
+    }
+
+    /// Initializes balance for specified address in any denominator.
+    fn init_balance_denom(env: &mut TestEnv, address: &Addr, amount: u128, denom: &str) {
+        env.router
+            .bank
+            .init_balance(&mut env.storage, address, coins(amount, denom))
             .unwrap();
+    }
 
-        (api, store, router, block, validator_addr)
+    /// Utility function for checking multiple balances in staking denominator.
+    fn assert_balances(env: &TestEnv, balances: impl IntoIterator<Item = (Addr, u128)>) {
+        for (addr, amount) in balances {
+            let balance: BalanceResponse = query_bank(
+                env,
+                BankQuery::Balance {
+                    address: addr.to_string(),
+                    denom: BONDED_DENOM.to_string(),
+                },
+            )
+            .unwrap();
+            assert_eq!(balance.amount.amount.u128(), amount);
+        }
     }
 
     #[test]
     fn add_get_validators() {
-        let api = MockApi::default();
-        let mut store = MockStorage::new();
-        let stake = StakeKeeper::default();
-        let block = mock_env().block;
+        let mut env = TestEnv::new((10, 100, 1), (0, 20, 1));
 
-        let validator_addr = api.addr_make("test-validator");
+        let validator_addr_3 = env.validator_addr_3();
```

Comment:
```suggestion
        let validator_addr_3 = api.addr_make("valoper3");
```

or extract the method `valoper_addr` and reuse here.

Reply:
It's not so obvious now, but this will be crucial for properly implementing this: #173.